src/Workspaces/Core/Portable/Options/GlobalOptionService.cs

```diff
@@ -39,44 +39,48 @@ internal sealed class GlobalOptionService(
-
-        // Local functions
-        static ImmutableArray<IOptionPersister> GetOptionPersistersSlow(
-            IWorkspaceThreadingService? workspaceThreadingService,
-            ImmutableArray<Lazy<IOptionPersisterProvider>> persisterProviders,
-            CancellationToken cancellationToken)
+        var result = GetOptionPersistersImplAsync(CancellationToken.None);
+        if (workspaceThreadingService is not null)
         {
-            if (workspaceThreadingService is not null && workspaceThreadingService.IsOnMainThread)
-            {
-                // speedometer tests report jtf.run calls from background threads, so we try to avoid those.
-                return workspaceThreadingService.Run(() => GetOptionPersistersAsync(persisterProviders, cancellationToken));
-            }
-            else
-            {
-                return GetOptionPersistersAsync(persisterProviders, cancellationToken).WaitAndGetResult_CanCallOnBackground(cancellationToken);
-            }
+            // Synchronous codepath through GetOption has potential JTF.Run invocation if it's the first option requested.
+            return workspaceThreadingService.Run(() => result);
+        }
+        else
+        {
+            return result.WaitAndGetResult_CanCallOnBackground(CancellationToken.None);
         }
+    }
 
-        static async Task<ImmutableArray<IOptionPersister>> GetOptionPersistersAsync(
-            ImmutableArray<Lazy<IOptionPersisterProvider>> persisterProviders,
-            CancellationToken cancellationToken)
+    private ValueTask<ImmutableArray<IOptionPersister>> GetOptionPersistersAsync(CancellationToken cancellationToken)
+    {
+        if (!_lazyOptionPersisters.IsDefault)
         {
-            return await persisterProviders.SelectAsArrayAsync(
-                static (lazyProvider, cancellationToken) => lazyProvider.Value.GetOrCreatePersisterAsync(cancellationToken),
-                cancellationToken).ConfigureAwait(false);
+            return new ValueTask<ImmutableArray<IOptionPersister>>(_lazyOptionPersisters);
         }
+
+        return new ValueTask<ImmutableArray<IOptionPersister>>(GetOptionPersistersImplAsync(cancellationToken));
+    }
+
+    private async Task<ImmutableArray<IOptionPersister>> GetOptionPersistersImplAsync(CancellationToken cancellationToken)
+    {
+        // Option persisters cannot be initialized while holding the global options lock
+        // https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1353715
+        Debug.Assert(!Monitor.IsEntered(_gate));
+
+        var lazyOptionPersisters = await _optionPersisterProviders.SelectAsArrayAsync(
+            static (lazyProvider, cancellationToken) => lazyProvider.Value.GetOrCreatePersisterAsync(cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        ImmutableInterlocked.InterlockedInitialize(
+            ref _lazyOptionPersisters,
+            lazyOptionPersisters);
+
+        return lazyOptionPersisters;
```

Comment:
is this correct?  or should you be grabbing _lazyOptionPErsisters?

Reply:
Yeah, _lazyOptionPersisters is probably better to use